### ISSUE TITLE
Maven :report-aggregate support nesting > 1

### ISSUE THREAD

**Issue by ls-urs-keller** [jacoco/jacoco#1241]
THIS IS A BUG TRACKER ONLY. FOR QUESTIONS PLEASE CHECK FAQ OR USE FORUM:

http://www.jacoco.org/jacoco/trunk/doc/faq.html

https://groups.google.com/forum/?fromgroups=#!forum/jacoco

Please understand that
ISSUES WITHOUT FOLLOWING INFORMATION WILL BE CLOSED WITHOUT COMMENTS!
Thank you for filling feature request!

### Scenario

* JaCoCo version: 0.8.7
* Operating system: All
* Tool integration: Maven 
* Description of your use case: (detailed description or executable reproducer, e.g. GitHub repo)

### Current Behaviour
Assume we have the following project structure:
- commons/pom.xml
- commons/src/main/xyz/Util.java
- sub-reactor/service-commons/pom.xml
- sub-reactor/service/pom.xml
- sub-reactor/pom.xml
- pom.xml

Assume a test in sub-reactor/service that covers a class xyz/Util.java in commons, which is a transitive dependency of service-commons, then the coverage report in sub-reactor/service will have no coverage for xyz/Util.java.

### Wanted Behaviour
An aggregated coverage report should also resolve transitive dependencies within the reactor. Currently only 1 level of dependencies for each project is resolved against the reactor.

### Possible Workarounds
A workaround would be to configure a report project and enumerate all dependencies.

**Comment by ls-urs-keller** [comment]
Suggested pull request, however not sure I understand the original intention of this:
```	/**
	 * Note that if dependency specified using version range and reactor
	 * contains multiple modules with same artifactId and groupId but of
	 * different versions, then first dependency which matches range will be
	 * selected. For example in case of range <code>[0,2]</code> if version 1 is
	 * before version 2 in reactor, then version 1 will be selected.
	 */
```

The project in the reactor will always have a fixed version, no range.
The dependency might have a version range. If that dependency points to the same version as present in the reactor then all is good.
Are there cases where a dependency should be covered but points to a version not in the reactor? In that case I can only see the following case:
- checked out project at version 1
- the project has also version 2 and it is published to maven
- in project version 1, a dependency range to a dependency within the reactor is used e.g. [1,100)
- this would mean that even when building 1 the project will take version 2 for a dependency, since the maven would resolve the newer version
- this would then mean that we are trying to report coverage from version 2 of the dep to the source of version 1 

Not sure if my analysis is correct and whether this makes sense.

**Comment by JayDi85** [comment]
It’s a very useful feature - use of maven’s reactor to collect all dependencies of multi-modules project. It’s hard to manually research and add all modules dependencies to the aggregation project (there are possible hundreds of deps).

Jacoco must collect all dependencies from listed modules by default without any hacks or workaround like https://github.com/jacoco/jacoco/pull/1251#issuecomment-1061585625

**Comment by tircis** [comment]
Hi,
Since I was struggling with same problem (Jacoco plugin doesn't following maven's reactor to collect its dependencies, in particular with multi-modules project), and reading a bunch of articles, i decided to fix it on my own with a dedicated maven-plugin that mimics Jacoco official one. @JayDi85 you may have a look at it and give me some feedback about it : [jacoco-aggregate-submodule](https://github.com/codefilarete/jacoco-aggregate-submodule)

**Comment by wingice** [comment]
Yes. It'd better to use maven's reactor to collect all dependencies of multi-modules project instead of creating a new module to hold all the dependent-modules.  Because we have a very large project with 30+ modules.

**Comment by rubenboadana** [comment]
Any plans to include support to this feature in a near future? 

Having to have yet another module to gather the reports in a huge multi-module project has more downsides than benefits imo, and having to define all the transitive dependencies as direct dependencies is a bit ugly and, in cases like ours, where we do sanity checks to avoid unused dependencies in our pom, does not work at all.

I see that it's a topic raised a couple of times and it never went forward, but I truly think that at least add it behind a flag would help a lot.

**Comment by yeroc** [comment]
Also wondering whether there are plans to add the feature as described by @JayDi85 and implemented by @tircis?  Would the team accept a PR that implements this either as a separate Mojo (perhaps `report-aggregate-modules`) or an option on the existing `ReportAggregateMojo`?  I'm also working with a fairly large mono-repo and it's not feasible to create a synthetic module with dependencies on everything.  Much easier and simpler to have a report execution in the top-level pom that runs after all tests have executed and aggregates across all sub-modules.
